File: typescript/01-que-es-un-agente/agente-react.ts
// Variante ReAct: mismo loop que agente-minimo, pero con CoT explicito (Thought antes de Action).

// Cómo ejecutar: make ts SCRIPT=typescript/01-que-es-un-agente/agente-react.ts
// Qué esperar: trace con Thought/Action/Observation antes de la respuesta final.

import Anthropic from "@anthropic-ai/sdk";

const MODEL = process.env["MODEL"] ?? "claude-sonnet-4-6";
const MAX_ITERATIONS = 10;

const SYSTEM =
  "Eres un agente ReAct. Antes de cada llamada a herramienta escribe una linea " +
  "que empiece por 'Thought:' explicando tu razonamiento; luego usa la herramienta. " +
  "Cuando tengas la respuesta final, escribela despues de un 'Final answer:'.";

const TOOLS: Anthropic.Tool[] = [
  {
    name: "get_time",
    description: "Devuelve la hora actual en una zona horaria (offset UTC en horas).",
    input_schema: {
      type: "object",
      properties: { utc_offset: { type: "number" } },
      required: ["utc_offset"],
    },
  },
  {
    name: "add",
    description: "Suma dos numeros.",
    input_schema: {
      type: "object",
      properties: { a: { type: "number" }, b: { type: "number" } },
      required: ["a", "b"],
    },
  },
];

function executeTool(name: string, args: Record<string, number>): string {
  if (name === "get_time") {
    const ms = Date.now() + args.utc_offset * 3600 * 1000;
    return new Date(ms).toISOString();
  }
  if (name === "add") return String(args.a + args.b);
  return `Tool '${name}' no existe`;
}

async function runReact(task: string): Promise<string> {
  const client = new Anthropic();
  const messages: Anthropic.MessageParam[] = [{ role: "user", content: task }];
  const trace: string[] = [];

  for (let i = 0; i < MAX_ITERATIONS; i++) {
    const response = await client.messages.create({
      model: MODEL,
      max_tokens: 1024,
      system: SYSTEM,
      tools: TOOLS,
      messages,
    });

    for (const block of response.content) {
      if (block.type === "text" && block.text.trim()) {
        trace.push(block.text.trim());
      } else if (block.type === "tool_use") {
        trace.push(`Action: ${block.name}(${JSON.stringify(block.input)})`);
      }
    }

    if (response.stop_reason === "end_turn" || response.stop_reason === "stop_sequence") {
      return trace.join("\n");
    }

    if (response.stop_reason === "tool_use") {
      const toolResults = response.content
        .filter((b): b is Anthropic.ToolUseBlock => b.type === "tool_use")
        .map((b) => {
          const out = executeTool(b.name, b.input as Record<string, number>);
          trace.push(`Observation: ${out}`);
          return { type: "tool_result" as const, tool_use_id: b.id, content: out };
        });
      messages.push({ role: "assistant", content: response.content });
      messages.push({ role: "user", content: toolResults });
      continue;
    }

    break;
  }

  return [...trace, "[max iteraciones]"].join("\n");
}

runReact("Que hora es en Tokio (UTC+9), y cuanto es 47 + 89?").then(console.log);

File: python/01-que-es-un-agente/agente-react.py
"""
Agente ReAct — loop con razonamiento explicito (Thought → Action → Observation).

Qué demuestra:
    ReAct (Reasoning + Acting) obliga al modelo a externalizar su razonamiento
    como texto visible antes de cada tool call. Util para depuracion y para
    modelos que razonan mejor escribiendo "Thought:" explicitamente.
    Diferencia con agente-minimo: el trace muestra pensamiento + acciones.

Patron clave:
    System prompt instruye formato Thought/Action/Observation.
    El codigo captura texto de cada turno y lo muestra en el trace.
    stop_reason == "end_turn" con "Final answer:" indica conclusion.

Cómo ejecutar:
    make py SCRIPT=python/01-que-es-un-agente/agente-react.py

Qué esperar:
    Trace de razonamiento + acciones antes de la respuesta final.
    Mas verbose que agente-minimo, pero el proceso es inspeccionable.

Variables de entorno:
    MODEL — modelo a usar (default: claude-sonnet-4-6)
"""
from datetime import datetime, timezone, timedelta
import json
import re
import os
import anthropic

MODEL = os.environ.get("MODEL", "claude-sonnet-4-6")
MAX_ITERATIONS = 10

SYSTEM = (
    "Eres un agente ReAct. Antes de cada llamada a herramienta escribe una linea "
    "que empiece por 'Thought:' explicando tu razonamiento; luego usa la herramienta. "
    "Cuando tengas la respuesta final, escribela despues de un 'Final answer:'."
)

# Algunos modelos locales (ej: llama3.1 via Ollama) escriben las acciones como
# texto JSON en vez de usar tool_use blocks. Este parser los detecta y ejecuta.
def _parse_text_actions(text: str) -> list[dict]:
    """Extrae llamadas a herramientas descritas como JSON en el texto del modelo."""
    decoder = json.JSONDecoder()
    calls = []
    i = 0
    while i < len(text):
        i = text.find("{", i)
        if i == -1:
            break
        try:
            obj, end = decoder.raw_decode(text, i)
            if isinstance(obj, dict):
                name = obj.get("name") or obj.get("tool")
                args = obj.get("parameters") or obj.get("arguments") or obj.get("input") or {}
                if name and isinstance(args, dict):
                    calls.append({"name": name, "input": args})
            i = end
        except (json.JSONDecodeError, ValueError):
            i += 1
    return calls

TOOLS = [
    {
        "name": "get_time",
        "description": "Devuelve la hora actual en una zona horaria (offset UTC en horas).",
        "input_schema": {
            "type": "object",
            "properties": {"utc_offset": {"type": "number"}},
            "required": ["utc_offset"],
        },
    },
    {
        "name": "add",
        "description": "Suma dos numeros.",
        "input_schema": {
            "type": "object",
            "properties": {"a": {"type": "number"}, "b": {"type": "number"}},
            "required": ["a", "b"],
        },
    },
]


def execute_tool(name: str, args: dict) -> str:
    if name == "get_time":
        tz = timezone(timedelta(hours=args["utc_offset"]))
        return datetime.now(tz=tz).isoformat()
    if name == "add":
        return str(args["a"] + args["b"])
    return f"Tool '{name}' no existe"


def run_react(task: str) -> str:
    client = anthropic.Anthropic()
    messages = [{"role": "user", "content": task}]
    trace: list[str] = []

    for _ in range(MAX_ITERATIONS):
        response = client.messages.create(
            model=MODEL,
            max_tokens=1024,
            system=SYSTEM,
            tools=TOOLS,
            messages=messages,
        )

        for block in response.content:
            if block.type == "text" and block.text.strip():
                trace.append(block.text.strip())
            elif block.type == "tool_use":
                trace.append(f"Action: {block.name}({block.input})")

        if response.stop_reason == "tool_use":
            tool_results = []
            for block in response.content:
                if block.type == "tool_use":
                    out = execute_tool(block.name, block.input)
                    trace.append(f"Observation: {out}")
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": block.id,
                        "content": out,
                    })
            messages.append({"role": "assistant", "content": response.content})
            messages.append({"role": "user", "content": tool_results})
            continue

        if response.stop_reason in ("end_turn", "stop_sequence"):
            # Modelos locales a veces escriben acciones como texto JSON en vez de
            # usar tool_use blocks. Si encontramos acciones en el texto, las ejecutamos.
            full_text = " ".join(b.text for b in response.content if b.type == "text")
            text_actions = _parse_text_actions(full_text)
            if text_actions:
                obs_parts = []
                for call in text_actions:
                    out = execute_tool(call["name"], call["input"])
                    trace.append(f"Action: {call['name']}({call['input']})")
                    trace.append(f"Observation: {out}")
                    obs_parts.append(f"{call['name']} → {out}")
                messages.append({"role": "assistant", "content": full_text})
                messages.append({"role": "user", "content": "Resultados: " + "; ".join(obs_parts) + ". Ahora da la respuesta final."})
                continue
            return "\n".join(trace)

        break

    return "\n".join(trace + ["[max iteraciones]"])


if __name__ == "__main__":
    print(run_react("Que hora es en Tokio (UTC+9), y cuanto es 47 + 89?"))
